apollo-federation/src/operation/merging.rs

```diff
@@ -173,119 +156,118 @@ impl SelectionSet {
+                        inline.merge_into(other_inline)?;
                     }
                 },
-                selection_map::Entry::Vacant(vacant) => {
-                    vacant.insert(other_selection.clone())?;
-                }
             }
+
+            Ok(())
         }
 
-        for (key, self_selection) in target.iter_mut() {
-            match self_selection {
-                SelectionValue::Field(mut self_field_selection) => {
-                    if let Some(other_field_selections) = fields.shift_remove(key) {
-                        self_field_selection.merge_into(
-                            other_field_selections.iter().map(|selection| &***selection),
-                        )?;
-                    }
-                }
-                SelectionValue::FragmentSpread(mut self_fragment_spread_selection) => {
-                    if let Some(other_fragment_spread_selections) =
-                        fragment_spreads.shift_remove(key)
-                    {
-                        self_fragment_spread_selection.merge_into(
-                            other_fragment_spread_selections
-                                .iter()
-                                .map(|selection| &***selection),
-                        )?;
-                    }
-                }
-                SelectionValue::InlineFragment(mut self_inline_fragment_selection) => {
-                    if let Some(other_inline_fragment_selections) =
-                        inline_fragments.shift_remove(key)
-                    {
-                        self_inline_fragment_selection.merge_into(
-                            other_inline_fragment_selections
-                                .iter()
-                                .map(|selection| &***selection),
-                        )?;
+        let target = Arc::make_mut(&mut self.selections);
+
+        if let Some(parental_directives) = directives_of_parent_selection {
+            fn recurse_on_inline_fragment<'a>(
+                target: &mut SelectionMap,
+                directives: &DirectiveList,
+                type_pos: &CompositeTypeDefinitionPosition,
+                others: impl Iterator<Item = &'a Selection>,
+            ) -> Result<(), FederationError> {
+                for selection in others {
+                    if let Selection::InlineFragment(inline) = selection {
+                        let inline_directives = &inline.inline_fragment.data().directives;
+                        if (inline_directives.len() == 0 || inline_directives == directives)
+                            && inline
+                                .inline_fragment
+                                .data()
+                                .type_condition_position
+                                .as_ref()
+                                .map(|pos| pos == type_pos)
+                                .unwrap_or(true)
+                        {
```

Comment:
```suggestion
                        if inline.is_unnecessary(&type_pos)
```

let's just keep it simple and do what JS does (no need to pass the parent directives as we don't have access to those when adding op path elements)

Reply:
Hmm... I see what you're saying. I was under the impression that `add_at_path` (or rather its call sites) did checks to see if the directives matched. I could easily be wrong, though.